### PR TITLE
Copy APIs for S3 and fs

### PR DESCRIPTION
**Pull request by prabhat393**
- Added [copyObject](https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3#S3.CopyObject) s3 API wrapper and Copy interface
- Tested s3 wrapper using minio
- Added Copy method for fs.

### CODE REVIEW

File: lib/fs/storage.go [hunk 1]
```diff
@@ -91,6 +92,32 @@ func (s Storage) WalkWithContext(_ context.Context, path string, callback func(p
 	return s.Walk(path, callback)
 }
 
+// Copy copies a file.
+func (s Storage) Copy(src string, dst string, options ...map[string]interface{}) error {
+	mode := 0644
+
+	if len(options) > 0 {
+		if options[0]["mode"] != nil {
```
Comment (protsack-stephan) [inline lib/fs/storage.go]: We can omit `if` statements, and just a for loop here. Will be little bit less conditions. Also we would probably do a lookup in a the map, making sure that it's `nil` is not the same thing as checking if key in the map exists. Please double check type casting as well to omit errors. Something like:
```go
for _, opt := range options {
	if v, ok := opt["mode"]; ok {
		if m, ok := v.(int); ok {
			mode = m
		}
	}
}
```

File: lib/fs/storage.go [hunk 2]
```diff
@@ -91,6 +92,32 @@ func (s Storage) WalkWithContext(_ context.Context, path string, callback func(p
...
+	}
+
+	input, err := ioutil.ReadFile(src)
+	if err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(dst, input, fs.FileMode(mode)); err != nil {
+		return err
+	}
```
Comment (protsack-stephan) [inline lib/fs/storage.go]: Please add empty line below to separate conditional from return statement.

File: lib/s3/storage.go [hunk 3]
```diff
@@ -171,6 +172,54 @@ func (s *Storage) WalkWithContext(ctx aws.Context, path string, callback func(pa
 	return nil
 }
 
+// Copy copies an object from the a path in a bucket to another path in the same or different bucket.
+// 'src' and 'dst' are absolute paths of the file.
+func (s *Storage) Copy(src string, dst string, options ...map[string]interface{}) error {
+	dstBucket := s.bucket
+
+	if len(options) > 0 {
+		if options[0]["dstBucket"] != nil {
```
Comment (protsack-stephan) [inline lib/s3/storage.go]: Same here, can be replaced with for loop.

File: lib/s3/storage.go [hunk 4]
```diff
@@ -171,6 +172,54 @@ func (s *Storage) WalkWithContext(ctx aws.Context, path string, callback func(pa
...
+		}
+	}
+
+	_, err := s.s3.CopyObject(&s3.CopyObjectInput{
+		Bucket:     aws.String(dstBucket),
+		CopySource: aws.String(fmt.Sprintf("%s/%s", s.bucket, src)),
+		Key:        aws.String(dst),
+	})
+
+	if err != nil {
```
Comment (protsack-stephan) [inline lib/s3/storage.go]: This if statement is redundant, we can just return the error.

File: lib/s3/storage.go [hunk 5]
```diff
@@ -171,6 +172,54 @@ func (s *Storage) WalkWithContext(ctx aws.Context, path string, callback func(pa
...
+}
+
+// CopyWithContext copies an object from the a path in a bucket to another path in the same or different bucket.
+// 'src' and 'dst' are absolute paths of the file.
+func (s *Storage) CopyWithContext(ctx aws.Context, src string, dst string, options ...map[string]interface{}) error {
+	dstBucket := s.bucket
+
+	if len(options) > 0 {
+		if options[0]["dstBucket"] != nil {
+			dstBucket = options[0]["dstBucket"].(string)
```
Comment (protsack-stephan) [inline lib/s3/storage.go]: Same here, can be replaces with the for loop.

File: lib/s3/storage.go [hunk 6]
```diff
@@ -171,6 +172,54 @@ func (s *Storage) WalkWithContext(ctx aws.Context, path string, callback func(pa
...
+		}
+	}
+
+	_, err := s.s3.CopyObjectWithContext(ctx, &s3.CopyObjectInput{
+		Bucket:     aws.String(dstBucket),
+		CopySource: aws.String(fmt.Sprintf("%s/%s", s.bucket, src)),
+		Key:        aws.String(dst),
+	})
+
+	if err != nil {
```
Comment (protsack-stephan) [inline lib/s3/storage.go]: This `if` statement is redundant, we can just return the error.

**Review by protsack-stephan — COMMENTED** [review]
Good work! Couple of comments.

File: lib/fs/storage.go [hunk 7]
```diff
@@ -91,6 +92,35 @@ func (s Storage) WalkWithContext(_ context.Context, path string, callback func(p
...
+
+	return nil
```
Comment (memije) [inline lib/fs/storage.go]: You can simplify this by returning the result of `ioutil.WriteFile`.

```suggestion
	return ioutil.WriteFile(dst, input, fs.FileMode(mode))
```

**Review by memije — COMMENTED** [review]
Looks good. Left just one small comment

**Pull request merged**